### PR TITLE
ACS-4064: fix for linking from a folder with inherited rules

### PR DESCRIPTION
**Pull request by gevagel**
This PR introduces changes aiming to fix the bug described in this ticket [https://alfresco.atlassian.net/browse/ACS-4064](https://alfresco.atlassian.net/browse/ACS-4064).

### CODE REVIEW

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/RuleSetLinksTests.java [hunk 1]
```diff
@@ -172,6 +173,56 @@ public void linkToRuleSet()
...
+    /**
+     * Check we can link to a rule set when linking from a folder which has inherited rules.
+     */
+    @Test(groups = {TestGroup.REST_API, TestGroup.RULES})
+    public void linkFromFolderWithInheritedRules()
+    {
+        STEP("Create folders");
+        final FolderModel parentFolder = dataContent.usingUser(user).usingSite(site).createFolder();
+        final FolderModel childFolder = dataContent.usingUser(user).usingResource(parentFolder).createFolder();
+        final FolderModel linkingFolder = dataContent.usingUser(user).usingSite(site).createFolder();
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/RuleSetLinksTests.java:185]:
```suggestion
        final FolderModel linkedToFolder = dataContent.usingUser(user).usingSite(site).createFolder();
```

File: repository/src/main/java/org/alfresco/repo/rule/RuleServiceImpl.java [hunk 2]
```diff
@@ -459,7 +459,13 @@ public boolean isRuleTypeEnabled(String ruleType)
     public boolean hasRules(NodeRef nodeRef)
     {
         return getRules(nodeRef).size() != 0;
-    } 
+    }
+
+    @Override
+    public boolean hasNonInheritedRules(NodeRef nodeRef)
```
Comment (KikoUK) [inline repository/src/main/java/org/alfresco/repo/rule/RuleServiceImpl.java:465]: Silly question: How does adding this method fix the issue?

**Review by mpichura — COMMENTED** [review]
Can you please verify if this issue does not happen in Share? Jamal is not certain this is an unwanted behavior. We should be like for like with Share.

File: repository/src/main/java/org/alfresco/service/cmr/rule/RuleService.java [hunk 3]
```diff
@@ -165,6 +165,15 @@
     @Auditable(parameters = {"nodeRef"})
     public boolean hasRules(NodeRef nodeRef);
 
+    /**
+     * Indicates whether the node in question has any non-inherited rules associated with it.
+     *
+     * @param nodeRef   the node reference
+     * @return          true if the node has rules associated, false otherwise
+     */
+    @Auditable(parameters = {"nodeRef"})
```
Comment (tpage-alfresco) [inline repository/src/main/java/org/alfresco/service/cmr/rule/RuleService.java:174]: `@Experimental`?

**Comment by tpage-alfresco** [comment]
> Can you please verify if this issue does not happen in Share? Jamal is not certain this is an unwanted behavior. We should be like for like with Share.

This is unwanted behaviour. Share certainly allows nodes that link to a rule set and inherit other rule sets.  I don't know for sure if it allows this sequence of events, but I'm pretty sure that it does.

File: repository/src/main/java/org/alfresco/repo/rule/RuleServiceImpl.java (hunk 2 above)
Comment (gevagel) [reply repository/src/main/java/org/alfresco/repo/rule/RuleServiceImpl.java:465]: Inside the method that does the actual linking `linkToRuleSet`, we are checking whether the folder that is trying to link has any pre-existing rules. With this method, we will ignore inherited rules in that check and perform the linking

**Comment by TomikaArome** [comment]
> Can you please verify if this issue does not happen in Share? Jamal is not certain this is an unwanted behavior. We should be like for like with Share.

I can confirm Share allows you to link a rule set to a folder with inherited rules already. I checked that before raising the bug

**Review by tpage-alfresco — APPROVED** [review]
LGTM

**Review by krdabrowski — APPROVED** [review]

**Pull request merged**